vllm/distributed/kv_transfer/kv_connector/utils.py

```diff
@@ -201,16 +202,36 @@ def update_finished_set(
                         aggregated_kv_connector_stats.aggregate(kv_connector_stats)
                     )
 
+            # Combine kv_cache_events from all workers.
+            if combined_kv_cache_events is None:
+                # Use the first worker's kv_cache events as start event list.
+                combined_kv_cache_events = kv_output.kv_cache_events
+            elif kv_cache_events := kv_output.kv_cache_events:
+                assert isinstance(
+                    combined_kv_cache_events,
+                    type(kv_cache_events),
+                )
+                worker_kv_cache_events = kv_cache_events.get_all_events()
+                combined_kv_cache_events.add_events(worker_kv_cache_events)
+
             invalid_block_ids |= kv_output.invalid_block_ids
 
         # select output of the worker specified by output_rank
         output = outputs[output_rank]
 
+        # Aggregate the events across workers.
+        # This operation needs to be done post worker processing so that we have all
+        # events for all workers.
+        if combined_kv_cache_events is not None:
+            combined_kv_cache_events.set_workers(self._expected_finished_count)
+            combined_kv_cache_events = combined_kv_cache_events.aggregate()
```

Comment:
`combined_kv_cache_events` holds up information on worker events on this step.
Events that did not accumulate for enough workers will be lost after this step.
You need to keep track of those events cross-steps.
See for example how they do it for `finished_recving` (using `self._recv_remaining_count`).

Reply:
That's interesting, thanks for letting me know and explaining this on slack.

I have pushed 717702fe5d6a41c6607ffe8c5c2736674e47ca86 which persists events across steps which is requied if workers are not processed in the same step. It then resets the events once the scheduler retrieves them in `take_events()`.